remove unused feature types in nnet3 scripts

LvHang opened the pull request:
Hi Dan,
I checked all the files in "egs/*/local/nnet3", "egs/*/local/chain" and "steps/nnet3" directories to make sure none of them set the "--feat-type" option.
I deleted the 'lda' branch in "steps/nnet3/ scripts" and fixed the codes about "transform_dir".  Now, the scripts only execute the feat-transform about "raw".
But there are three exceptions.
(1)"steps/nnet3/chain/build_tree.sh" looks only support "delta" feature. (2)"egs/wsj/s5/steps/nnet3/get_degs.sh" and "steps/nnet3/make_denlats.sh" support "raw" and "delta". 
In the comment of "steps/nnet3/make_denlats.sh", it said "you can set this in order to run on top of delta features, although we don't normally want to do this." So I keep it. I don't know if I should delete it.

Now, I'm testing it. When the testing finish, I will inform you.
Hang

danpovey commented on egs/wsj/s5/steps/nnet3/chain/build_tree.sh:
```diff
@@ -85,19 +85,10 @@ echo $nj >$dir/num_jobs
-
-if [ -f $alidir/final.mat ]; then feat_type=lda; else feat_type=delta; fi
-echo "$0: feature type is $feat_type"
+echo "$0: feature type is delta"
```
sorry-- you should actually leave this file as it was.  it uses the features from a baseline GMM-based system, which will normally be LDA.

danpovey commented on egs/wsj/s5/steps/nnet3/decode.sh:
```diff
@@ -86,33 +85,18 @@ echo $nj > $dir/num_jobs
-    echo "$0: LDA transforms differ between $srcdir and $transform_dir"
-    exit 1;
-  fi
+  trans=raw_trans;
```
no need to use a variable now since it can have only one value.

danpovey commented on egs/wsj/s5/steps/nnet3/decode_looped.sh:
```diff
@@ -94,33 +93,18 @@ echo $nj > $dir/num_jobs
-    echo "$0: LDA transforms differ between $srcdir and $transform_dir"
-    exit 1;
-  fi
+  trans=raw_trans;
```
same here, remove the variable.

danpovey commented on egs/wsj/s5/steps/nnet3/get_degs.sh:
```diff
@@ -13,7 +13,7 @@ cmd=run.pl
 
 # feature options
-feat_type=raw     # set it to 'lda' to use LDA features.
+feat_type=raw     # set it to 'delta' to use delta features.
```
disable delta features here; this only runs on top of already-trained chain system so if we remove delta in the basic chain training scripts (which you should), it has no reason to exist here.

danpovey commented on egs/wsj/s5/steps/nnet3/get_egs_targets.sh:
```diff
@@ -132,14 +129,7 @@ fi
-  fi
-fi
-if [ -f $transform_dir/raw_trans.1 ] && [ $feat_type == "raw" ]; then
+if [ -f $transform_dir/raw_trans.1 ]; then
```
@vimalmanohar, are we ever using get_egs_target.sh with other than raw features?

danpovey commented on egs/wsj/s5/steps/nnet3/lstm/train.sh:
```diff
@@ -103,7 +103,6 @@ egs_opts=
 transform_dir=     # If supplied, this dir used instead of alidir to find transforms.
 cmvn_opts=  # will be passed to get_lda.sh and get_egs.sh, if supplied.
```
No need to change this script, it is deprecated so I don't want to waste time testing the change.

danpovey commented on egs/wsj/s5/steps/nnet3/make_denlats.sh:
```diff
@@ -34,8 +34,8 @@ extra_left_context=0
 extra_left_context_initial=-1
 extra_right_context_final=-1
-feat_type=  # you can set this in order to run on top of delta features, although we don't
-            # normally want to do this.
```
No need to have this variable any more.  This script doesn't need to support delta features or any features exept raw, and there is no need for a "feat_type" variable.

LvHang replied:
Weird, I updated it but it isn't changed. Let me see.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train.sh:
```diff
@@ -71,7 +71,6 @@ egs_opts=
 transform_dir=     # If supplied, this dir used instead of alidir to find transforms.
 cmvn_opts=  # will be passed to get_lda.sh and get_egs.sh, if supplied.
             # only relevant for "raw" features, not lda.
```
you can revert changes to this script, the script is deprecated.

danpovey commented on egs/wsj/s5/steps/nnet3/train_tdnn.sh:
```diff
@@ -70,7 +70,6 @@ egs_opts=
 transform_dir=     # If supplied, this dir used instead of alidir to find transforms.
 cmvn_opts=  # will be passed to get_lda.sh and get_egs.sh, if supplied.
             # only relevant for "raw" features, not lda.
-feat_type=raw  # or set to 'lda' to use LDA features.
```
you can also revert changes to this script.

LvHang commented:
Hi Dan,
I have already addressed these comments.
Today, I can write a discriminative training example script refer to swbd's relevant script, then I will push it. After that, @vimalmanohar  or  @hhadian can help to review it. Thanks for your work.
And I will test it with mini_librispeech setup.
Hang

LvHang commented:
@vimalmanohar @hhadian 
I have added a discriminative recipe for mini_librispeech. Could you please help to review it when you are free. Thanks.
I'm testing it. I write it refer to the discriminative scripts of swbd and librispeech. I hope there is no essential error.
Hang

danpovey commented on egs/mini_librispeech/s5/local/chain/tuning/run_tdnn_1c_discriminative.sh:
```diff
@@ -0,0 +1,221 @@
+frames_overlap_per_eg=30
+
+## Nnet training options
+effective_learning_rate=0.000001
```
if you copied this from a setup with more data, it may make sense to increase it.
but with this little data, you may not get any improvement at all so it may be hard to tune this value.

LvHang replied:
Yeah, I refer to some scripts for the "effective_learning_rate" option.
In tedlium, it is 1.25*10^-6. In librispeech, it is 1*10^-6. In swbd, it is 1.25*10^-7.
But they are all much bigger than mini_librispeech.
I don't have enough experience for it. So do you have any suggestive value for this option?

danpovey commented:
Try 10 times larger.

On Sun, Jun 25, 2017 at 1:51 AM, LvHang <notifications@github.com> wrote:

> *@LvHang* commented on this pull request.
> ------------------------------
>
> In egs/mini_librispeech/s5/local/chain/tuning/run_tdnn_1c_
> discriminative.sh
> <https://github.com/kaldi-asr/kaldi/pull/1711#discussion_r123890931>:
>
> > +tree_dir=exp/chain/tree_sp
> +degs_dir=                     # If provided, will skip the degs directory creation
> +lats_dir=                     # If provided, will skip denlats creation
> +
> +## Objective options
> +criterion=smbr
> +one_silence_class=true
> +
> +dir=${srcdir}_${criterion}
> +
> +## Egs options
> +frames_per_eg=150
> +frames_overlap_per_eg=30
> +
> +## Nnet training options
> +effective_learning_rate=0.000001
>
> Yeah, I refer to some scripts for the "effective_learning_rate" option.
> In tedlium, it is 1.25*10^-6. In librispeech, it is 1*10^-6. In swbd, it
> is 1.25*10^-7.
> But they are all much bigger than mini_librispeech.
> I don't have enough experience for it. So do you have any suggestive value
> for this option?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#discussion_r123890931>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu97I3ybtL-hJ-axnjRJXsRKomfebks5sHfVsgaJpZM4OEN1Z>
> .
>

LvHang commented:
Ok, thanks.

LvHang commented:
I check the steps/nnet3/train_discriminative.sh, it doesn't contain the "--adjust-priors" and  "--modify-learning-rates" options. 
But the discriminative scripts of swbd（swbd/s5c/local/chain/tuning/run_tdnn_6h_discriminative.sh） and librispeech（librispeech/s5/local/chain/run_tdnn_discriminative.sh ） still use them.
So I guess they are out-of-date, right?
Hang

danpovey commented:
@vimalmanohar, any idea what is the problem here?


On Sun, Jun 25, 2017 at 10:59 PM, LvHang <notifications@github.com> wrote:

> I check the steps/nnet3/train_discriminative.sh, it doesn't contain the
> "--adjust-priors" and "--modify-learning-rates" options.
> But the discriminative scripts of swbd（swbd/s5c/local/chain/
> tuning/run_tdnn_6h_discriminative.sh） and librispeech（librispeech/s5/
> local/chain/run_tdnn_discriminative.sh ） still use them.
> So I guess they are out-of-date, right?
> Hang
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310952641>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu8DhFVYsPobN4SpiVOIvDLO6drcjks5sHx6OgaJpZM4OEN1Z>
> .
>

vimalmanohar commented:
adjust-priors was made true by default and the option is removed. Also
modify-learning-rates is no longer required and is not supported anymore.


On Sun, Jun 25, 2017 at 11:02 PM Daniel Povey <notifications@github.com>
wrote:

> @vimalmanohar, any idea what is the problem here?
>
>
> On Sun, Jun 25, 2017 at 10:59 PM, LvHang <notifications@github.com> wrote:
>
> > I check the steps/nnet3/train_discriminative.sh, it doesn't contain the
> > "--adjust-priors" and "--modify-learning-rates" options.
> > But the discriminative scripts of swbd（swbd/s5c/local/chain/
> > tuning/run_tdnn_6h_discriminative.sh） and librispeech（librispeech/s5/
> > local/chain/run_tdnn_discriminative.sh ） still use them.
> > So I guess they are out-of-date, right?
> > Hang
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310952641>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ADJVu8DhFVYsPobN4SpiVOIvDLO6drcjks5sHx6OgaJpZM4OEN1Z
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310952966>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEATV2QbH1AzD2dxQ-vlwXADqc-kGT5lks5sHx8ugaJpZM4OEN1Z>
> .
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University

danpovey commented:
OK,  @HangLv, please search through all example scripts for the use of
those options and remove them.  (obviously excluding nnet2 scripts).


On Sun, Jun 25, 2017 at 11:12 PM, Vimal Manohar <notifications@github.com>
wrote:

> adjust-priors was made true by default and the option is removed. Also
> modify-learning-rates is no longer required and is not supported anymore.
>
>
> On Sun, Jun 25, 2017 at 11:02 PM Daniel Povey <notifications@github.com>
> wrote:
>
> > @vimalmanohar, any idea what is the problem here?
> >
> >
> > On Sun, Jun 25, 2017 at 10:59 PM, LvHang <notifications@github.com>
> wrote:
> >
> > > I check the steps/nnet3/train_discriminative.sh, it doesn't contain
> the
> > > "--adjust-priors" and "--modify-learning-rates" options.
> > > But the discriminative scripts of swbd（swbd/s5c/local/chain/
> > > tuning/run_tdnn_6h_discriminative.sh） and librispeech（librispeech/s5/
> > > local/chain/run_tdnn_discriminative.sh ） still use them.
> > > So I guess they are out-of-date, right?
> > > Hang
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310952641>,
> > or mute
> > > the thread
> > > <
> > https://github.com/notifications/unsubscribe-auth/
> ADJVu8DhFVYsPobN4SpiVOIvDLO6drcjks5sHx6OgaJpZM4OEN1Z
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310952966>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AEATV2QbH1AzD2dxQ-
> vlwXADqc-kGT5lks5sHx8ugaJpZM4OEN1Z>
> > .
> >
> --
> Vimal Manohar
> PhD Student
> Electrical & Computer Engineering
> Johns Hopkins University
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-310954177>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu0-z0xede2mjT92sOnEUn2heKyIAks5sHyGigaJpZM4OEN1Z>
> .
>

LvHang commented:
Yes, i see.

LvHang commented:
@danpovey @vimalmanohar 
I think when we use steps/nnet3/decode.sh to decode the discriminative model, the script will call "steps/nnet2/check_ivectors_compatible.sh" to check the ivectors compatible between "ivector-dir" and "discriminative dir". But we have never copied the "final.ie.id" of the base model dir(eg. exp/chain/tdnn1c_sp). So that it always prompt warning, right?
Hang

danpovey commented:
Probably the discriminative-training script should copy that-- please
include that in your PR.  But make sure the script doesn't fail if the
final.ie.id is not present in the source directory.  If the script does set
-e at the top, then you may have to do something like
    cp $srcdir/final.ie.id $dir 2>/dev/null || true
to suppress any error.


On Mon, Jun 26, 2017 at 5:32 PM, LvHang <notifications@github.com> wrote:

> @danpovey <https://github.com/danpovey> @vimalmanohar
> <https://github.com/vimalmanohar>
> I think when we use steps/nnet3/decode.sh to decode the discriminative
> model, the script will call "steps/nnet2/check_ivectors_compatible.sh" to
> check the ivectors compatible between "ivector-dir" and "discriminative
> dir". But we have never copied the "final.ie.id" of the base model
> dir(eg. exp/chain/tdnn1c_sp). So that it always prompt warning, right?
> Hang
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-311188383>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu9A2DyFN37GvUTI6REFCZSu8JBQDks5sICOHgaJpZM4OEN1Z>
> .
>

LvHang commented:
@danpovey @vimalmanohar 
I think there is also a mistake. Now the adjust-priors was made true by default. so that the "steps/nnet3/adjust_prior.sh" always will be called.  This script will generate the `${dir}/${iter}_adj.mdl`. However, in swbd corpus, when it run "steps/nnet3/decode.sh", the "--iter epoch$x.adj" will be set, that means it will use `${dir}/${iter}.adj.mdl` model. Obviously, the name is wrong. I will fix it in my PR.
For a question, when we do decoding in discriminative script, which kind of model we always use by default, the "adj.mdl" or ".mdl"?
Hang

danpovey commented:
I think you should try decoding both of them and see which is better.  I
don't have a lot of experience with the discriminative training and I don't
recall which is normally better.  Probably the adjusted one, but I'm not
sure.

On Mon, Jun 26, 2017 at 7:56 PM, LvHang <notifications@github.com> wrote:

> @danpovey <https://github.com/danpovey> @vimalmanohar
> <https://github.com/vimalmanohar>
> I think there is also a mistake. Now the adjust-priors was made true by
> default. so that the "steps/nnet3/adjust_prior.sh" always will be called.
> This script will generate the ${dir}/${iter}_adj.mdl. However, in swbd
> corpus, when it run "steps/nnet3/decode.sh", the "--iter epoch$x.adj" will
> be set, that means it will use ${dir}/${iter}.adj.mdl model. Obviously,
> the name is wrong. I will fix it in my PR.
> For a question, when we do decoding in discriminative script, which kind
> of model we always use by default, the "adj.mdl" or ".mdl"?
> Hang
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/1711#issuecomment-311213911>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu7uyR5AAdf1EPpxsi5U15jAj3tk4ks5sIEUZgaJpZM4OEN1Z>
> .
>

LvHang commented:
Hi Dan,
This is the newest results about discriminative training. (offline-decoding)
	                tglarge_dev_clean_2	tgsmall_dev_clean_2
baseline	                10.85	                            14.9
epoch1	                10.16	                            14.3
epoch1_adj	        10.13	                            14.26
epoch2	               *10.11	                            14.33
epoch2_adj	         10.21	                          *14.23
epoch3	                 10.41	                            14.49
epoch3_adj	         10.36	                            14.37
epoch4	                 10.53	                            14.56
epoch4_adj	         10.5	                            14.49

The "_adj"  means I use the "adjust_prior" model. The one without "_adj" means the normal model.
The numerical value with a asterisk in front of it means the best result.

Hang

LvHang commented:
Hi Dan,
This is the newest results about discriminative training. (offline-decoding)
···
                      tglarge_dev_clean_2	tgsmall_dev_clean_2
baseline	                10.85	                  14.9
epoch1	                10.16	                  14.3
epoch1_adj	         10.13	                 14.26
epoch2	                *10.11	                 14.33
epoch2_adj	          10.21	                *14.23
epoch3	                  10.41	                 14.49
epoch3_adj	          10.36	                 14.37
epoch4	                  10.53	                 14.56
epoch4_adj	          10.5	                 14.49
···
The "_adj" means I use the "adjust_prior" model. The one without "_adj" means the normal model.
The numerical value with a asterisk in front of it means the best result.

Hang

LvHang commented:
Hi Dan,
This is the newest results about discriminative training. (offline-decoding)

|                   | tglarge_dev_clean_2	| tgsmall_dev_clean_2 |
|-------------|------------------------|------------------------|
|baseline	    |         10.85	                |           14.9 | 
|epoch1	    |         10.16	                |           14.3 |
|epoch1_adj |	     10.13	                |           14.26 |
|epoch2	    |       *10.11	                |           14.33 |
|epoch2_adj |	     10.21	                |          *14.23 |
|epoch3	    |         10.41	                |           14.49 |
|epoch3_adj |	     10.36	                |           14.37 |
|epoch4	    |         10.53	                |           14.56 |
|epoch4_adj |	     10.5	                |           14.49 |

The "_adj" means I use the "adjust_prior" model. The one without "_adj" means the normal model.
The numerical value with a asterisk in front of it means the best result.

Hang

danpovey commented on egs/wsj/s5/steps/nnet3/align.sh:
```diff
@@ -78,38 +77,20 @@ cp $lang/phones.txt $dir || exit 1;
-    echo "$0: LDA transforms differ between $srcdir and $transform_dir"
-    exit 1;
-  fi
+  trans=raw_trans;
```
please don't use a variable here, just replace `$trans` with raw_trans

LvHang commented:
I have fixed it. Sorry for the carelessness.

danpovey commented:
merging-- thanks.
